### ISSUE TITLE
controlling download access to uploaded files.

### ISSUE THREAD

**Issue by wesyah234** [tomitrescak/meteor-uploads#125]
I am loving this uploader tool so far, but I've come to a snag, which was discussed on issue 58: #58 

I have previously tried doing file uploads with CollectionFS, but was looking for something better.  Specifically I liked the idea that I could choose my own collection in which to store my file metadata, which didn't seem to be possible with the CollectionFS way of doing uploads.

So I put together a demo (https://github.com/wesyah234/fileUploadDemo2) using this package, and soon came to realize that controlling access to the uploaded files did not seem to be possible, at least in the easy way it was with CollectionFS.  See https://github.com/CollectionFS/Meteor-CollectionFS#security for how they are doing it, via allow/deny rules.

I was able to prevent a view of a file depending on whether the user was logged in, because it hooked into the Meteor.Accounts system.

The solution in issue #58 does not seem like it will tie into the Meteor.accounts system or anything like that.  Are there any other ideas for this issue?

In my work with Java and PHP, typically you would stream the document through PHP or Java, and before streaming it through, you would get to check certain permissions.  Is there a possible solution like that here with Meteor?  IE. stream it through a route in Meteor?

**Comment by tomitrescak** [comment]
I have good news for you. All you have to do, is to modify the behaviour of the upload-server. Check out the code here for the GET part of the server: https://github.com/tomitrescak/meteor-tomi-upload-server/blob/master/upload_server.js#L202-L239

If you want to control access to individual uploaded files, you can add access token to file URL and then on server you would only validate that token.

**Comment by wesyah234** [comment]
I'm not sure I understand why I would have to modify your upload_server.js code...

Couldn't I just use the validateRequest function? it seems that's the intention of that method.  Seems like I could check some token in the url request, and then return false to prevent the view of the file.  But... I tried that and it still downloads, even though I'm returning false:
````
 Meteor.startup(function () {
    UploadServer.init({
      tmpDir: process.env.PWD + '/Uploads/tmp',
      uploadDir: process.env.PWD + '/Uploads/',
      checkCreateDirectories: true,
      finished: function (fileInfo, formFields) {
        console.log("upload finished, fileInfo ", fileInfo);
        console.log("upload finished, formFields: ", formFields);
        Uploads.insert(fileInfo);
      },
      cacheTime: 100,
      mimeTypes: {
        "xml": "application/xml",
        "vcf": "text/x-vcard"
      },
      validateRequest: function (req, res) {
        console.log("validate request: ", req);
        console.log("token passed in: ", req.query.token);
        console.log("validating request, rerurning false");
        return false;
//        console.log("validate reqeust, resp: ", res);
      }
    })
  });
````

**Comment by tomitrescak** [comment]
Ha! That is absolutely correct .. I apologise. This change was done via pull request and I forgot about it. Thank god for careful programmers like you.

**Comment by tomitrescak** [comment]
According to following code, you have to return error in validateRequest:

```javascript
var error = options.validateRequest(req, res);
    if (error) {
      res.writeHead(403, {'Content-Type': 'text/plain'});
      res.write(error);
      res.end();
      return;
    }
```

sou you would do

```javascript
validateRequest: function (req, res) {
        console.log("validate request: ", req);
        console.log("token passed in: ", req.query.token);
        console.log("validating request, rerurning false");
        return "You cheeky!! No files for you!";
//        console.log("validate reqeust, resp: ", res);
}
```

**Comment by wesyah234** [comment]
ok, that's good, I'll try that... but now, what I really want to do is know whether the user accessing the file is authenticated or not through Meteor.accounts... do you have any suggestions for dealing with that?

**Comment by wesyah234** [comment]
What I'm thinking is this:  on the client, I can have a helper that returns:
````
Meteor.user().services.resume.loginTokens[0].hashedToken
````
and all the url download links that I generate can have ?token=(that token above from helper)

and in this validateRequest function I can do this:
````
      validateRequest: function (req, res) {
        console.log("token passed in: ", req.query.token);
        var user = Users.findOne({"services.resume.loginTokens":{$elemMatch:{"hashedToken":req.query.token}}});
        console.log("user ", user);
      }
````
And, if that findOne query gives me a valid user document, then I know they are probably logged in, or at least they were logged in recently, which is better than nothing.

if the find one fails, I can return the string to cause the download to fail.

From looking at the loginTokens in my Users collection, it looks like the tokens go away after about a day.

**Comment by wesyah234** [comment]
Update: I can improve things on the client by getting not the oldest, but the newest loginToken by this:
```
Meteor.user().services.resume.loginTokens[Meteor.user().services.resume.loginTokens.length-1].hashedToken
```
However, I'm concerned about a couple of things:
1. if I build download links for my files and include h=398djfdkjf9309jfjdfjdf9j, am I exposing the user's login token when I shouldn't be?
2. If the user closes the browser, but leaks the download url to someone else, the url will continue to work for a day or so, until the Meteor or mongo system gets around to cleaning out those resume tokens from the collection. (I still don't know how they go away, but they have consistently gone away after a day)

Comparing it to the https://github.com/CollectionFS/Meteor-CollectionFS  package, I see the "links" generated for the download files contain a "token" but that token does not seem to be my resume token in the Users collection, so I'm not sure where they get that from...

Bottom line, I love your front end download UI, and I also love having more control over where I store the file metadata, but I also love how the collectionFS package lets me control the downloads based on whether the user is logged in or not.  If only these 2 packages could merge somehow :)

**Comment by tomitrescak** [comment]
Ha! I thought I responded .. I did .. but via email and it did not land here. I apologise. Did you solve it? I'll be soon building the authentication plugin for the files. IO have some ideas, but not sure when that will happen. Till then, we are bound to use tokens ... if used correctly .. you should not be worried to exposing them to client. You can incorporate incorporate timestamps with them, limiting them time wise.

**Comment by wesyah234** [comment]
The solution I went with was to add a "prepare download" button , and the user clicks that, which will generate a timestamp and a token and write to the database.  Then, reactively, the prepare download button will change to a "download" link, which will only be valid for a period of time.  I will probably use 30 seconds or so.

**Comment by wesyah234** [comment]
The server side will then check the token on the database, and verity the time has not expired, and allow or not allow the download.
